documentation/api/query/v4/pql.markdown

```diff
@@ -0,0 +1,361 @@
+---
+title: "PuppetDB 3.2 » API » v4 » Puppet query language (PQL)"
+layout: default
+canonical: "/puppetdb/latest/api/query/v4/pql.html"
+---
+
+[entities]: ./entities.html
+[subquery]: #subqueries
+[ast]: ./ast.html
+
+> **Experimental Feature**: This featureset is experimental, and it may be altered or removed in
+> future release.
+
+Puppet Query Language (PQL) is a query language designed with PuppetDB and Puppet data in mind. It
+provides a string based query language as an alternative to the [AST query language][ast] PuppetDB has
+always supported.
+
+## Query Structure
+
+A PQL query has the following structure:
+
+    <entity> [<projection>] { <filter> <modifiers> }
+
+Which is broken up into the following parts:
+
+* `entity`: Rquired. The entity context that this query executes on.
+* `projection`: Optional. Modifies the output of how the `entity` fields are returned.
+* `filter`: Optional. The filter to match on for this `entity`.
+* `modifiers`: Optional. Contains modifiers for the query.
+
+As an example, if you wanted to query for `nodes` related data, but only see the `certnames` field
+for each node, with a regex filter across certname:
+
+    nodes[certname] { certname ~ '^web' }
+
+In this case, this would return only the certname field of nodes starting with `web`.
+
+## Entity
+
+The entity or context of a query (or subquery) defines what results you will get returned when
+performing a query, and provides the main context for any projections or filters in the query.
+There are many entities, for a full list see the [entities] documentation.
+
+For PQL queries, the entity context is the minimal amount of information one must provide, as it
+defines the results returned. For example, if you wanted to see all node information, you could
+provide a query as follows:
+
+    nodes {}
+
+And it would be enough to return all node data, without filtering or pagination.
+
+The entity context can also be used within a subquery, see the [subquery] section for more details.
+
+## Projection
```

Comment:
@mullr so, I have thought about it re: bare entities. I kind of feel like a bare empty would feel a bit unintuitive, and originally I was considering always ensuring entities had some other marker. Puppet uses a capitalized convention in some parts (ie. Resources, Nodes) but for resource pointers ... for example ... and that made me thing using that might make more sense, then everything that is capitalized is definitely an entity, not something else like a keyword. Outside of Puppet though, this is pretty weird.

One point - An empty extract is quite common, an empty filter is very uncommon I would presume, at least for a lot of people. Even those who want the house, will want to paginate for example ... and right now the pagination marker is planned to be within the {} (I guess, I'm ... so unsure about this though).

We could go the other way, and require nothing to be optional, and that makes some symmetry:

```nodes[]{}```

Kind of looks ugly though :-). Maybe the extraction should be mandatory, filter optional:

```nodes[]```

Hohum.